front/src/applications/stdcm/components/StdcmResults/StdcmResults.tsx

```diff
@@ -124,26 +124,68 @@ const StdcmResults = ({
         };
       });
 
-      const request = {
+      const baseRequest = {
         infra_id: infraId,
-        rolling_stock: {
-          name: consist.tractionEngine?.name ?? '',
-          speed_limit_tag: consist.speedLimitByTag,
-        },
         timetable_id: timetableId,
         waypoints,
       };
 
-      const response = await postSimilarTrains({ body: request });
-      const rawSimilarTrains = response.data?.similar_trains ?? [];
+      const hasNullTrains = (trains: PostSimilarTrainsApiResponse['similar_trains']) =>
+        trains.some((segment) => segment.train === null);
+
+      // Call 1: Full criteria
+      const fullCriteriaResponse = await postSimilarTrains({
+        body: {
+          ...baseRequest,
+          rolling_stock: {
+            name: consist.tractionEngine?.name ?? '',
+            speed_limit_tag: consist.speedLimitByTag,
+          },
+        },
+      });
+
+      const fullCriteriaSegments = fullCriteriaResponse.data?.similar_trains ?? [];
+
+      // Call 2: Only speed_limit_tag (if needed)
+      const speedTagResponseSegments =
+        hasNullTrains(fullCriteriaSegments) && consist.speedLimitByTag
+          ? ((
+              await postSimilarTrains({
+                body: {
+                  ...baseRequest,
+                  rolling_stock: { speed_limit_tag: consist.speedLimitByTag },
+                },
+              })
+            ).data?.similar_trains ?? [])
+          : [];
+
+      // Call 3: Only name (if still needed)
+      const nameOnlySegments =
+        hasNullTrains(speedTagResponseSegments) && consist.tractionEngine?.name
+          ? ((
+              await postSimilarTrains({
+                body: {
+                  ...baseRequest,
+                  rolling_stock: { name: consist.tractionEngine?.name ?? '' },
```

Comment:
You know now that the name is defined
```suggestion
                  rolling_stock: { name: consist.tractionEngine.name },
```